### PR TITLE
Adding pipeline scripts for build and deploy.

### PR DESCRIPTION
**Pull request by Nkuruba**
Creating a pull request to test the pipeline build and deploy operations. There may be more commits to be added to this pull request.

### CODE REVIEW

File: .pipeline/lib/clean.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  const options = settings.options
+  const oc=new OpenShiftClientX(Object.assign({'namespace':phases.build.namespace}, options));
+  const target_phase=options.env
+
+  for (var k in phases){
+    if (phases.hasOwnProperty(k)) {
+      const phase=phases[k]
+      if (k == target_phase ){
+        //console.log(`phase=${phase}`)
+        //oc.raw('get', ['bc'], {selector:`app-name=${phase.name},env-id=${phase.changeId},env-name!=prod,!shared,github-repo=${oc.git.repository},github-owner=${oc.git.owner}`, namespace:phase.namespace, 'output':'custom-columns=kind:.spec.output.to.kind,name:.spec.output.to.name', 'no-headers':'true'})
```
Comment (andrea-williams) [inline .pipeline/lib/clean.js]: will we ever need this commented-out code again or can we just delete it?

File: .pipeline/lib/config.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+'use strict';
+const options= require('pipeline-cli').Util.parseArguments()
+const changeId = options.pr //aka pull-request
+const version = '1.0.0'
+const name = 'wps-web'
+
+Object.assign(options.git, {owner: 'bcgoc', repository: 'wps-web'})
```
Comment (andrea-williams) [inline .pipeline/lib/config.js]:
```suggestion
Object.assign(options.git, {owner: 'bcgov', repository: 'wps-web'})
```

File: openshift/deployment.yaml [hunk 3]
```diff
@@ -48,15 +48,16 @@ objects:
...
+        app: ${NAME}${SUFFIX}
     template:
       metadata:
         labels:
           app: ${NAME}${SUFFIX}
       spec:
         automountServiceAccountToken: false
         containers:
-          - image: docker-registry.default.svc:5000/auzhsi-tools/wps-web:latest
+          #- image: docker-registry.default.svc:5000/auzhsi-tools/wps-web:latest
```
Comment (andrea-williams) [inline openshift/deployment.yaml]: should this still be in here?

**Comment by DerekRoberts** [comment]
@andrea-williams @Kyubinhan Let's not worry too much about the JavaScript code in here.  It's from another project and planned to be replaced.  @Nkuruba is on it.

**Comment by DerekRoberts** [comment]
@Nkuruba It's not a blocker, but please don't forget to have Jenkins notify us only for TEST and PROD deployments.  Builds maybe only PROD.

**Comment by Nkuruba** [comment]
@DerekRoberts as of now I did not include TEST env. but for PROD, the notifications are being sent to our rocketchat channel.

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                        currentBuild.rawBuild.delete()
+                        error("No changes detected in the module path for ${MODULE_NAME} (${BUILD_TRIGGER_EXCLUDES})")
+                    }
+                }
+                echo "Aborting all running jobs for ${MODULE_NAME}"
+                script {
+                    abortAllPreviousBuildInProgress(currentBuild)
+                }
+                echo "Building ..."
+                sh "curl -X POST -H \"Content-Type: application/json\" https://chat.pathfinder.gov.bc.ca/hooks/7A8RwSpZPDAqc6uXS/z8nRXX43oifYWELkwJKeLj4e86rtyfvtJdqdvjSE3XoK2sri -d '{\"text\":\"Building wps-web for PR-${CHANGE_ID}.\"}'"
```
Comment (DerekRoberts) [inline Jenkinsfile]: Please limit Rocket.Chat notifications to TEST and PROD deployments.  Possibly only PROD for builds, if not less.

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                sh "cd .pipeline && chmod +x npmw && ./npmw ci DEBUG='*' && ./npmw run build -- --pr=${CHANGE_ID} --git.branch.name=${CHANGE_BRANCH} --git.branch.merge=${CHANGE_BRANCH} --git.branch.remote=${CHANGE_BRANCH} --git.url=${GIT_URL} --git.change.target=${CHANGE_TARGET}"
+            }
+        }
+        stage('Deploy (DEV)') {
+            agent { label 'deploy' }
+            steps {
+                echo "Deploying ..."
+                sh "cd .pipeline && chmod +x npmw && ./npmw ci && ./npmw run deploy -- --pr=${CHANGE_ID} --env=dev --git.branch.name=${CHANGE_BRANCH} --git.branch.merge=${CHANGE_BRANCH} --git.branch.remote=${CHANGE_BRANCH} --git.url=${GIT_URL} --git.change.target=${CHANGE_TARGET}"
```
Comment (DerekRoberts) [inline Jenkinsfile]: Please limit Rocket.Chat notifications to TEST and PROD deployments.  Possibly only PROD for builds, if not less.

**Comment by DerekRoberts** [comment]
> @DerekRoberts as of now I did not include TEST env. but for PROD, the notifications are being sent to our rocketchat channel.

@Nkuruba 
Deployment notices are still hitting DEV. 

```
Pipeline-Jenkins-wps-web @rocket.cat
Bot
2:00 PM
Deployed wps-web to DEV for PR-7.
```

**Comment by DerekRoberts** [comment]
> @DerekRoberts as of now I did not include TEST env. but for PROD, the notifications are being sent to our rocketchat channel.

@Nkuruba We've already had this discussion.  Please lighten up on the Rocket.Chat notifications, particularly DEV.

**Comment by Nkuruba** [comment]
@DerekRoberts removed notification for DEV deployment.

File: .pipeline/package.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  },
+  "scripts": {
+    "build": "node build.js",
+    "clean": "node clean.js",
+    "deploy": "node deploy.js",
+    "version": "echo \"node@$(node --version) ($(which node))\" && echo \"npm@$(npm --version) ($(which npm))\" && npm ls"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/bcgov/ocp-sso.git"
```
Comment (Sybrand) [inline .pipeline/package.json]: Shouldn't this be the repo of our project?

File: openshift/deployment.yaml [hunk 7]
```diff
@@ -8,30 +8,30 @@ parameters:
   - name: TAG
     value: latest
     description: The output image tag beig deployed
-#  - name: DOCKER_REGISTRY
-#    required: true
+  - name: DOCKER_REGISTRY
```
Comment (Sybrand) [inline openshift/deployment.yaml]: Are we using docker as a stop-gap?

File: .pipeline/npmw [hunk 8]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+set +x
+type -t nvm && nvm deactivate
+export NVM_DIR="$(git rev-parse --show-toplevel)/.nvm"
+if [ ! -f "$NVM_DIR/nvm.sh" ]; then
+  mkdir -p "${NVM_DIR}"
+  curl -sSL -o- https://raw.githubusercontent.com/creationix/nvm/v0.34.0/install.sh | bash &>/dev/null
```
Comment (Sybrand) [inline .pipeline/npmw]: Why are we installing nvm? If we're using one of the red hat images, aren't we constrained to whatever version of node they're running anyway?

File: openshift/build.yaml [hunk 9]
```diff
@@ -9,7 +9,7 @@ parameters:
   - name: GIT_URL
     value: https://github.com/bcgov/wps-web.git
   - name: GIT_REF
-    value: master
+    value: nkuruba-master
```
Comment (Sybrand) [inline openshift/build.yaml]: Isn't there a better way to point the template to the branch we want to use? Hard coding to a particular branch doesn't seem like a good solution.

Comment (Nkuruba) [reply openshift/build.yaml]: The branch name has been updated to master.

File: .pipeline/package.json (hunk 6 above)
Comment (Nkuruba) [reply .pipeline/package.json]: Thats the source repository of pipeline code.

File: openshift/deployment.yaml (hunk 7 above)
Comment (Nkuruba) [reply openshift/deployment.yaml]: That's just a parameter key and the value is being passed to it through Jenkins file.

File: .pipeline/npmw (hunk 8 above)
Comment (Nkuruba) [reply .pipeline/npmw]: nvm is being used to run the pipeline scripts and it has nothing to do with the application.

File: openshift/deployment.yaml (hunk 3 above)
Comment (Nkuruba) [reply openshift/deployment.yaml]: It will be removed in future changes/commits.

**Pull request merged**